src/compiler/checker.ts

```diff
@@ -442,22 +442,76 @@ module ts {
             return result;
         }
 
+        function isImportSymbolDeclaration(node: Node): boolean {
+            return node.kind === SyntaxKind.ImportEqualsDeclaration ||
+                node.kind === SyntaxKind.ImportClause && !!(<ImportClause>node).name ||
+                node.kind === SyntaxKind.NamespaceImport ||
+                node.kind === SyntaxKind.ImportSpecifier;
+        }
+
+        function getDeclarationOfImportSymbol(symbol: Symbol): Declaration {
+            return forEach(symbol.declarations, d => isImportSymbolDeclaration(d) ? d : undefined);
+        }
+
+        function getTargetOfImportEqualsDeclaration(node: ImportEqualsDeclaration): Symbol {
+            if (node.moduleReference.kind === SyntaxKind.ExternalModuleReference) {
+                var moduleSymbol = resolveExternalModuleName(node, getExternalModuleImportEqualsDeclarationExpression(node));
+                var exportAssignmentSymbol = moduleSymbol && getResolvedExportAssignmentSymbol(moduleSymbol);
+                return exportAssignmentSymbol || moduleSymbol;
+            }
+            return getSymbolOfPartOfRightHandSideOfImportEquals(<EntityName>node.moduleReference, node);
+        }
+
+        function getTargetOfImportClause(node: ImportClause): Symbol {
+            var moduleSymbol = resolveExternalModuleName(node, (<ImportDeclaration>node.parent).moduleSpecifier);
+            if (moduleSymbol) {
+                var exportAssignmentSymbol = getResolvedExportAssignmentSymbol(moduleSymbol);
+                if (!exportAssignmentSymbol) {
+                    error(node.name, Diagnostics.External_module_0_has_no_default_export_or_export_assignment, symbolToString(moduleSymbol));
+                }
+                return exportAssignmentSymbol;
+            }
+        }
+
+        function getTargetOfNamespaceImport(node: NamespaceImport): Symbol {
+            return resolveExternalModuleName(node, (<ImportDeclaration>node.parent.parent).moduleSpecifier);

```

Comment:
That makes sense, although I think importing a module with an empty namespace using a namespace import is likely not what the user intended. If they just wanted to execute the module, they would have done `import "mod";`.

The error that you get for using the default import with a non-default-exported module is early. It's on the import. However, the error for importing an empty namespace with a namespace import is downstream. I could envision a user not being able to access a particular property on their import alias, and then wasting time looking inside the external module to figure out what happened.